Model/Indexer/AbstractIndexer.php

```diff
@@ -0,0 +1,358 @@
+
+    /** @var Emulation */
+    private $storeEmulator;
+
+    /** @var InputInterface */
+    private $input;
+
+    /** @var IndexerStatusServiceInterface */
+    private $indexerStatusService;
+
+    /**
+     * AbstractIndexer constructor.
+     * @param NostoHelperAccount $nostoHelperAccount
+     * @param NostoHelperScope $nostoHelperScope
+     * @param NostoLogger $nostoLogger
+     * @param StoreDimensionProvider $dimensionProvider
+     * @param Emulation $storeEmulator
+     * @param InputInterface $input
+     * @param IndexerStatusServiceInterface $indexerStatusService
+     * @param ProcessManager|null $processManager
+     */
+    public function __construct(
+        NostoHelperAccount $nostoHelperAccount,
+        NostoHelperScope $nostoHelperScope,
+        NostoLogger $nostoLogger,
+        StoreDimensionProvider $dimensionProvider,
+        Emulation $storeEmulator,
+        InputInterface $input,
+        IndexerStatusServiceInterface $indexerStatusService,
+        ProcessManager $processManager = null
+    ) {
+        $this->nostoHelperAccount = $nostoHelperAccount;
+        $this->nostoHelperScope = $nostoHelperScope;
+        $this->nostoLogger = $nostoLogger;
+        $this->dimensionProvider = $dimensionProvider;
+        $this->processManager = $processManager;
+        $this->input = $input;
+        $this->storeEmulator = $storeEmulator;
+        $this->indexerStatusService = $indexerStatusService;
+    }
+
+    /**
+     * Get ModeSwitcher class to later get the indexer mode
+     *
+     * @return ModeSwitcherInterface
+     */
+    abstract public function getModeSwitcher(): ModeSwitcherInterface;
+
+    /**
+     * Implement logic of single store indexing
+     *
+     * @param Store $store
+     * @param array $ids
+     */
+    abstract public function doIndex(Store $store, array $ids = []);
+
+    /**
+     * @return string
+     */
+    abstract public function getIndexerId(): string;
+
+    /**
+     * @inheritdoc
+     * @throws Exception
+     */
+    public function executeFull()
+    {
+        if ($this->allowFullExecution() === true) {
+            $indexerId = $this->getIndexerId();
+            $message = sprintf(
+                'Begin a full reindex for indexer "%s"',
+                $indexerId
+            );
+            $this->nostoLogger->info($message);
+            $this->doWork();
+            $this->nostoLogger->info('Finished full reindex');
+        }
+    }
+
+    /**
+     * @inheritdoc
+     * @throws Exception
+     */
+    public function executeList(array $ids)
+    {
+        $this->execute($ids);
+    }
+
+    /**
+     * @inheritdoc
+     * @throws Exception
+     */
+    public function executeRow($id)
+    {
+        $indexerId = $this->getIndexerId();
+        $message = sprintf(
+            'Begin a row reindex for indexer "%s" for "%s"',
+            $indexerId,
+            $id
+        );
+        $this->nostoLogger->info($message);
+        $this->execute([$id]);
```

Comment:
Should this be calling `execute` or `executeList`?

Reply:
It should be `execute`. That's the one Magento calls when doing the indexing